File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.py
# DB 캐시를 적용할 때 캐시 크기에 따른 실행시간 측정 프로그램
# cacheSize : 캐시 크기 / cities : 도시 이름 배열 -> 영문자, 대소문자 구분 X  : 대소문자 체크
# cache hit 1 / cache miss 3
# 캐시 크기 => 기억하는 국가 개수, 가장 오래전에 참조된 항목 삭제
def solution(cacheSize, cities):
    answer = 0
    cache = dict()
    now_cache = dict()
    for i in range(len(cities)):
        if cities[i].lower() in now_cache:
            answer += 1
        else:
            answer += 5
            
        cache.setdefault(cities[i].lower(), 0)
        cache[cities[i].lower()] = i
        now_cache = dict(sorted(cache.items(), key = lambda x:x[1], reverse = True)[:cacheSize]).keys()

    return answer

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.java
// 영문자, 대소문자 구분 X  : 다 소문자로 저장
// 11 16 19 20
import java.util.*;
class Solution {
    public int solution(int cacheSize, String[] cities) {
        int answer = 0;
        ArrayList<String> pages = new ArrayList<>();
        for (String city:cities){
            String cityName = city.toLowerCase();
            // 포함 여부 체크 -> 지우고 맨 뒤로 보내기
            if(pages.contains(cityName)) {
                pages.remove(cityName);
                pages.add(cityName);
                answer += 1;
            } else {
                // 포함하지 않는다면 -> 캐시 사이즈를 채우지 않은 경우 / 채운 경우
                if (pages.size() < cacheSize) {
                    pages.add(cityName);
                    answer += 5;
                } else if (cacheSize == 0){
                    answer += 5;
                } else {
                    pages.remove(0);
                    pages.add(cityName);
                    answer += 5;
                }
            }
        }
        return answer;
    }
}